sphinxgallery/gen_rst.py

```diff
@@ -118,133 +79,134 @@ def flush(self):
+.. rst-class:: sphx-glr-script-out
+
+  ::
+
+    {0}\n"""
+
 
-def extract_docstring(filename, ignore_heading=False):
-    """ Extract a module-level docstring, if any
+def split_code_and_text_blocks(source_file):
+    """Return list with source file separated into code and text blocks.
+
+    Returns
+    -------
+    blocks : list of (label, content)
+        List where each element is a tuple with the label ('text' or 'code'),
+        and content string of block.
     """
-    lines = open(filename).readlines()
-    start_row = 0
-    if lines[0].startswith('#!'):
-        lines.pop(0)
-        start_row = 1
-    docstring = ''
-    first_par = ''
-    line_iterator = iter(lines)
-    tokens = tokenize.generate_tokens(lambda: next(line_iterator))
-    for tok_type, tok_content, _, (erow, _), _ in tokens:
-        tok_type = token.tok_name[tok_type]
-        if tok_type in ('NEWLINE', 'COMMENT', 'NL', 'INDENT', 'DEDENT'):
-            continue
-        elif tok_type == 'STRING':
-            docstring = eval(tok_content)
-            # If the docstring is formatted with several paragraphs, extract
-            # the first one:
-            paragraphs = '\n'.join(
-                line.rstrip() for line
-                in docstring.split('\n')).split('\n\n')
-            if paragraphs:
-                if ignore_heading:
-                    if len(paragraphs) > 1:
-                        first_par = re.sub('\n', ' ', paragraphs[1])
-                        first_par = ((first_par[:95] + '...')
-                                     if len(first_par) > 95 else first_par)
-                    else:
-                        raise ValueError("Docstring not found by gallery.\n"
-                                         "Please check the layout of your"
-                                         " example file:\n {}\n and make sure"
-                                         " it's correct".format(filename))
-                else:
-                    first_par = paragraphs[0]
-
-        break
-    return docstring, first_par, erow + 1 + start_row
-
-
-def extract_line_count(filename, target_dir):
-    """Extract the line count of a file"""
-    example_file = os.path.join(target_dir, filename)
-    lines = open(example_file).readlines()
-    start_row = 0
-    if lines and lines[0].startswith('#!'):
-        lines.pop(0)
-        start_row = 1
-    line_iterator = iter(lines)
-    tokens = tokenize.generate_tokens(lambda: next(line_iterator))
-    check_docstring = True
-    erow_docstring = 0
-    for tok_type, _, _, (erow, _), _ in tokens:
-        tok_type = token.tok_name[tok_type]
-        if tok_type in ('NEWLINE', 'COMMENT', 'NL', 'INDENT', 'DEDENT'):
+
+    with open(source_file) as f:
+        source_lines = f.read()
+
+    doc_string = ast.get_docstring(ast.parse(source_lines))
+
+    blocks = [('text', doc_string)]
+    # Remove from source_lines the docstring
+    end_doc = [m.end() for m in re.finditer('"""', source_lines)][1]
```

Comment:
This is a bit fragile.

Looking at the ast.get_docstring code, I came up with a function that takes a filename and returns the docstring + the rest of the content of the file (feel free to tweak it any way you see fit):

```python
def get_module_docstring_and_rest(filename):
    with open(filename) as f:
        content = f.read()

    node = ast.parse(content)
    if not isinstance(node, ast.Module):
        raise TypeError("This function only supports modules. You provided {0}".format(node.__class__.__name__))
    if node.body and isinstance(node.body[0], ast.Expr) and \
       isinstance(node.body[0].value, ast.Str):
        docstring_node = node.body[0]
        docstring = docstring_node.value.s
        rest = content.split('\n', maxsplit=docstring_node.lineno)[-1]
        return docstring, rest
    else:
        raise ValueError(('Could not find docstring in file "{0}". '
                          'A docstring is required by sphinx-gallery').format(filename))
```